### PR TITLE
Add a new baseline vgg16

### PR DESCRIPTION
**Pull request by cedricgsh**

### CODE REVIEW

File: torchvision_models/segmentation/segmentation.py [hunk 1]
```diff
@@ -149,3 +150,17 @@ def erfnet_resnet(pretrained_weights='erfnet_encoder_pretrained.pth.tar', num_cl
                 original_weights[my_key] = saved_weights[key]
         net.load_state_dict(original_weights)
     return net
+
+def vgg16(pretrained_weights='pytorch-pretrained', num_classes=19, aux=0,
+          dropout_1=0.1, dropout_2=0.3, flattened_size=4500, scnn=False):
```
Comment (voldemortX) [inline torchvision_models/segmentation/segmentation.py:155]: dropout?

File: main_landec_as_seg.py [hunk 2]
```diff
@@ -23,7 +22,9 @@
     parser.add_argument('--dataset', type=str, default='tusimple',
                         help='Train/Evaluate on TuSimple (voc) / CULane (culane) (default: tusimple)')
     parser.add_argument('--model', type=str, default='scnn',
-                        help='Model selection (baseline/scnn/sad) (default: scnn)')
+                        help='Model selection (scnn/sad/none) (default: scnn)')
+    parser.add_argument('--baseline', type=str, default='erfnet',
```
Comment (voldemortX) [inline main_landec_as_seg.py:26]: --backbone=vgg/erfnet
--method=baseline/scnn/sad/etc.

File: torchvision_models/segmentation/segmentation.py [hunk 3]
```diff
@@ -149,3 +150,17 @@ def erfnet_resnet(pretrained_weights='erfnet_encoder_pretrained.pth.tar', num_cl
...
+    """Different from erfnet.
```
Comment (voldemortX) [inline torchvision_models/segmentation/segmentation.py:156]: nxp这注释。。。明显没话找话

File: torchvision_models/segmentation/vgg16.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        output = self.linear1(output)
+        output = F.relu(output)
+        output = self.linear2(output)
+        if predict:
+            output = torch.sigmoid(output)
+        return output
+
+
+# SCNN head
+class SpatialConv(nn.Module):
```
Comment (voldemortX) [inline torchvision_models/segmentation/vgg16.py:31]: 这个根ERFNet一样吧，直接import得了

File: torchvision_models/segmentation/vgg16.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+        output = self.encoder(input)
+        output = self.fc67(output)
+
+        if self.scnn is not None:
+            output = self.scnn(output)
+
+        output = self.fc8(output)
+        out['out'] = output
+        output = self.act(output)
```
Comment (voldemortX) [inline torchvision_models/segmentation/vgg16.py:139]: 多余吧

File: torchvision_models/segmentation/vgg16.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        self.act = nn.Sequential(
+            nn.Softmax(dim=1)
+        )
+
+        if aux > 0:
+            self.aux_head = LaneExistVgg(num_output=aux, flattened_size=flattened_size)
+        else:
+            self.aux_head = None
+
+    def forward(self, input, only_encode=False):
```
Comment (voldemortX) [inline torchvision_models/segmentation/vgg16.py:128]: 不用的变量都删掉

File: torchvision_models/segmentation/vgg16.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        for i in [34, 37, 40]:
+            conv = self.net._modules[str(i)]
+            dilated_conv = nn.Conv2d(
+                conv.in_channels, conv.out_channels, conv.kernel_size, stride=conv.stride,
+                padding=tuple(p * 2 for p in conv.padding), dilation=2, bias=(conv.bias is not None)
+            )
+            dilated_conv.load_state_dict(conv.state_dict())
+            self.net._modules[str(i)] = dilated_conv
+        self.net._modules.pop('33')
+        self.net._modules.pop('43')
```
Comment (voldemortX) [inline torchvision_models/segmentation/vgg16.py:83]: 这个确定么，lua版本能看出来有没有这个pop操作么

File: main_landec_as_seg.py [hunk 8]
```diff
@@ -22,8 +21,10 @@
                         help='Validation frequency (default: 0), 0: no online evaluation')
     parser.add_argument('--dataset', type=str, default='tusimple',
                         help='Train/Evaluate on TuSimple (voc) / CULane (culane) (default: tusimple)')
-    parser.add_argument('--model', type=str, default='scnn',
-                        help='Model selection (baseline/scnn/sad) (default: scnn)')
+    parser.add_argument('--method', type=str, default='scnn',
```
Comment (voldemortX) [inline main_landec_as_seg.py:24]:
```suggestion
    parser.add_argument('--method', type=str, default='baseline',
```

File: LANEDETECTION.md [hunk 9]
```diff
@@ -34,35 +34,42 @@ python tools/tusimple_list_convertor.py
...
 ```
-python main_landec_as_seg.py --epochs=12 --lr=0.15 --batch-size=20 --dataset=culane --model=baseline --mixed-precision --exp-name=<whatever you like>
+python main_landec_as_seg.py --epochs=12 --lr=0.15 --batch-size=20 --dataset=culane --method=none --backbone=erfnet --mixed-precision --exp-name=<whatever you like>
 ```
 
 Mixed precision training on TuSimple with ERFNet:
 
 ```
-python main_landec_as_seg.py --epochs=12 --lr=0.7 --batch-size=20 --dataset=tusimple --model=baseline --mixed-precision --exp-name=<whatever you like>
+python main_landec_as_seg.py --epochs=12 --lr=0.7 --batch-size=20 --dataset=tusimple --method=none --backbone=erfnet --mixed-precision --exp-name=<whatever you like>
```
Comment (voldemortX) [inline LANEDETECTION.md:43]:
```suggestion
python main_landec_as_seg.py --epochs=12 --lr=0.7 --batch-size=20 --dataset=tusimple --method=baseline --backbone=erfnet --mixed-precision --exp-name=<whatever you like>
```

File: main_landec_as_seg.py [hunk 10]
```diff
@@ -55,17 +56,18 @@
...
-    scnn = True if args.model == 'scnn' else False
+    scnn = True if args.method == 'scnn' else False
     weights = torch.tensor(weights).to(device)
-    if args.dataset == 'tusimple':
+    if args.dataset == 'tusimple' and args.backbone == 'erfnet':
         net = erfnet_tusimple(num_classes=num_classes, scnn=scnn)
-    elif args.dataset == 'culane':
+    elif args.dataset == 'culane'and args.backbone == 'erfnet':
         net = erfnet_culane(num_classes=num_classes, scnn=scnn)
+    elif args.dataset == 'culane'and args.backbone == 'vgg16':
```
Comment (voldemortX) [inline main_landec_as_seg.py:65]:
```suggestion
    elif args.dataset == 'culane' and args.backbone == 'vgg16':
```

File: LANEDETECTION.md [hunk 11]
```diff
@@ -34,35 +34,42 @@ python tools/tusimple_list_convertor.py
 Mixed precision training on CULane with ERFNet:
 
 ```
-python main_landec_as_seg.py --epochs=12 --lr=0.15 --batch-size=20 --dataset=culane --model=baseline --mixed-precision --exp-name=<whatever you like>
+python main_landec_as_seg.py --epochs=12 --lr=0.15 --batch-size=20 --dataset=culane --method=none --backbone=erfnet --mixed-precision --exp-name=<whatever you like>
```
Comment (voldemortX) [inline LANEDETECTION.md:37]:
```suggestion
python main_landec_as_seg.py --epochs=12 --lr=0.15 --batch-size=20 --dataset=culane --method=baseline --backbone=erfnet --mixed-precision --exp-name=<whatever you like>
```

File: LANEDETECTION.md [hunk 12]
```diff
@@ -34,35 +34,42 @@ python tools/tusimple_list_convertor.py
...
 ```
 
 Mixed precision training on TuSimple with ERFNet-SCNN:
 
 ```
-python main_landec_as_seg.py --epochs=12 --lr=0.3 --batch-size=20 --dataset=tusimple --model=scnn --mixed-precision --exp-name=<whatever you like>
+python main_landec_as_seg.py --epochs=12 --lr=0.3 --batch-size=20 --dataset=tusimple --method=scnn --backbone=erfnet --mixed-precision --exp-name=<whatever you like>
 ```
 
+Mixed precision training on CULane with Vgg16-SCNN:
```
Comment (voldemortX) [inline LANEDETECTION.md:58]:
```suggestion
Mixed precision training on CULane with VGG-SCNN:
```

File: torchvision_models/segmentation/deeplab_vgg.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+        output = self.encoder(input)
+        output = self.fc67(output)
+
+        if self.scnn is not None:
+            output = self.scnn(output)
+
+        output = self.fc8(output)
+        out['out'] = output
+        output = self.softmax(output)
```
Comment (voldemortX) [inline torchvision_models/segmentation/deeplab_vgg.py:98]: softmax应该在if里面

File: torchvision_models/segmentation/deeplab_vgg.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        output = self.linear1(output)
+        output = F.relu(output)
+        output = self.linear2(output)
+        if predict:
+            output = torch.sigmoid(output)
+        return output
+
+
+
+class Vgg16(nn.Module):
```
Comment (voldemortX) [inline torchvision_models/segmentation/deeplab_vgg.py:30]: 还是这个叫VGG16，下面叫DeepLabV1吧
